src/rulesets/Base/array.jl

```diff
@@ -196,7 +199,9 @@ function rrule(::typeof(hvcat), rows, values::Union{AbstractArray, Number}...)
             end
             dY[index...]
         end
-        return (NoTangent(), NoTangent(), dXs...)
+        _call((f, x)) = f(x)
+        proj_dXs = _call.(tuple(zip(project_Vs, dXs)...)) # can't make it type infer, e.g. test_rrule(hvcat, 1, rand(3)', transpose(rand(3)) ⊢ rand(1,3))
+        return (NoTangent(), NoTangent(), proj_dXs...)
```

Comment:
@oxinabox any ideas here?

Reply:
I suspect it is failing to infer the length.
Which is like https://github.com/JuliaDiff/ChainRulesTestUtils.jl/issues/185#issuecomment-871715279
so see the solutions there.

I think maybe changing the dims arg to the rrule to 
```julia
, dims::Tuple{Vararg{Union{Colon,Int},N}}) where N
```
and then  as an assert:
```
proj_dXs::NTuple{N} = 
```
Or on the other side. as a convert

Or might need it right on the end after the return (which would be a convert)

Reply:
The length infers fine. It can't tell what the elements are, even though it can if it is doing them one at a time. This is the ambiguity when `@code_warntype` the `pb` of the example in the comment.
```julia
Body::Tuple{NoTangent, NoTangent, Union{Adjoint{Float64, Matrix{Float64}}, Transpose{Float64, Matrix{Float64}}}, Union{Adjoint{Float64, Matrix{Float64}}, Transpose{Float64, Matrix{Float64}}}}
```